### ISSUE TITLE
Cuda backend build failure

### ISSUE THREAD

**Issue by hunbssfy** [google/iree#6979]
**Describe the bug**
I am using `-DIREE_TARGET_BACKENDS_TO_BUILD="CUDA"` and `-DIREE_HAL_DRIVERS_TO_BUILD="Cuda"` and latest code (tag: snapshot-20210906.509) to build IREE but got a bunch of errors like the following:

> LLVM ERROR: Building op `gpu.thread_id` but it isn't registered in this MLIRContext
> PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
> make[2]: Leaving directory '/mnt/workspace/iree/holmes_out'
> LLVM ERROR: Building op `gpu.thread_id` but it isn't registered in this MLIRContext
> [ 97%] Built target iree_test_e2e_xla_ops_check_vmvx_vmvx_floor.mlir_module
> [ 97%] Built target iree_test_e2e_xla_ops_check_cuda_dynamic_slice.mlir_module
> LLVM ERROR: Building op `gpu.thread_id` but it isn't registered in this MLIRContext
> PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
> Stack dump:
> 0.Program arguments: ../../../tools/[ 97%] Built target iree_test_e2e_xla_ops_check_vmvx_vmvx_dynamic_update_slice.mlir_module
> iree-translate -iree-mlir-to-vm-bytecode-module -mlir-print-op-on-diagnostic=false --iree-hal-target-backends=cuda -iree-input-type=mhlo /mnt/workspace/iree/iree/iree/test/e2e/xla_ops/compare.mlir -o check_cuda_compare.mlir_module.vmfb -iree-llvm-embedded-linker-path=/mnt/workspace/iree/holmes_out/third_party/llvm-project/llvm/bin/lld
> 1.make[2]: Leaving directory '/mnt/workspace/iree/holmes_out'
> iree/test/e2e/xla_ops/CMakeFiles/iree_test_e2e_xla_ops_check_cuda_scatter.mlir_module.dir/build.make:62: recipe for target 'iree/test/e2e/xla_ops/check_cuda_scatter.mlir_module.vmfb' failed
> make[2]: *** [iree/test/e2e/xla_ops/check_cuda_scatter.mlir_module.vmfb] Aborted (core dumped)

**To Reproduce**
1. I configure the build like this: 
```
cmake -GNinja -B $outdir . \
    -DCMAKE_BUILD_TYPE=Debug \
    -DIREE_TARGET_BACKENDS_TO_BUILD="CUDA;VMVX;DYLIB-LLVM-AOT;WASM-LLVM-AOT" \
    -DIREE_HAL_DRIVERS_TO_BUILD="Cuda;VMVX;DyLib"
```
2. Run the build process: `cmake --build $outdir` and you can produce the error.

**And if I drop the 'CUDA' and 'Cuda' from Backends and drivers config, the build passes.**
And I found that by adding "Vulkan-SPIRV" as a backend target and "Vulkan" as a driver, the build passes

**Comment by GMNGeoffrey** [comment]
That's weird. We register the GPU dialect here: https://github.com/google/iree/blob/1d65e9a79a7c8d63c3012d58dbad4960f6a6f02e/iree/tools/init_mlir_dialects.h#L41

This looks suspiciously like https://llvm.discourse.group/t/linalg-intermediate-ir-error/4251 where Alex explains that the complaint about having failed to register an op is actually an indication of having failed to *load* a dialect. So I think there's 

Part of the issue here is an ergonomics one where we're running the compiler at build time so the debugging is a little harder to do. That's my fault. But I think the issue is some pass (probably in MLIR core actually) not declaring that it's dependent on the GPU dialect (creates ops from that dialect). Lemme see if I can run iree-translate directly to figure out which pass

**Comment by GMNGeoffrey** [comment]
Ok I was able to reproduce the error you got. Except I also got a stack trace

https://gist.github.com/GMNGeoffrey/884dbeffec89b685d31db3c3df3674c2

The issue is in [LLVMGPUTileAndDistribute](https://github.com/google/iree/blob/main/iree/compiler/Codegen/LLVMGPU/LLVMGPUTileAndDistribute.cpp), 
which is weird because it does declare GPU as a dependent dialect: 

https://github.com/google/iree/blob/1d65e9a79a7c8d63c3012d58dbad4960f6a6f02e/iree/compiler/Codegen/LLVMGPU/LLVMGPUTileAndDistribute.cpp#L199-L201

With all those lambdas, the stack trace is a bit hard to follow, but I'm pretty sure this is firing when the pass is actually running, so the GPU dialect should be registered

PS. Also not sure why your log says stuff about `make` if you were using Ninja :-)

**Comment by GMNGeoffrey** [comment]
Based on that other thread, I'm suspicious that this is a new bug from upstream

**Comment by GMNGeoffrey** [comment]
@ThomasRaoux I thought that the CI was building the CUDA stuff as part of the turing build, but it looks like all it's setting is `IREE_CUDA_DISABLE=0` which just controls the driver filter, so the driver and backends aren't really built at all:

https://github.com/google/iree/blob/main/build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh

**Comment by GMNGeoffrey** [comment]
Going a week back this is still broken, so it's not just some new update from LLVM

**Comment by hunbssfy** [comment]
> With all those lambdas, the stack trace is a bit hard to follow, but I'm pretty sure this is firing when the pass is actually running, so the GPU dialect should be registered
 

Sorry, this was from a build where I forgot to assign `-GNinja` ...

**Comment by ThomasRaoux** [comment]
> @ThomasRaoux I thought that the CI was building the CUDA stuff as part of the turing build, but it looks like all it's setting is `IREE_CUDA_DISABLE=0` which just controls the driver filter, so the driver and backends aren't really built at all:
> 
> https://github.com/google/iree/blob/main/build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-turing/build.sh

`export IREE_CUDA_DISABLE=0` means CUDA is on and it should build CUDA HAL. Note that this test is part of the compiler which is built and tested anyway on all of our CI.

The default build should definitely runs this test.

**Comment by ThomasRaoux** [comment]
> Sorry, this was from a build where I was assiging it the wrong way : `--GNinja`

So this fixes the problem for you?

**Comment by hunbssfy** [comment]
> > Sorry, this was from a build where I was assiging it the wrong way : `--GNinja`
> 
> So this fixes the problem for you?

No, it doesn't fix the problem. `Ninja` build also fails as Geoffrey has reproduced the error.

**Comment by ThomasRaoux** [comment]
> > > Sorry, this was from a build where I was assiging it the wrong way : `--GNinja`
> > 
> > 
> > So this fixes the problem for you?
> 
> No, it doesn't fix the problem. `Ninja` build also fails as Geoffrey has reproduced the error.

Sounds good let me look at it.

> Note that this test is part of the compiler which is built and tested anyway on all of our CI.
Scratch that part, I read to fast this is xla test. it is ran on Turing and passes though:
`328/762 Test #249: iree/test/e2e/xla_ops/check_cuda_scatter.mlir ...............................................................   Passed    3.48 sec`

https://source.cloud.google.com/results/invocations/65260d52-b718-49b5-bd0a-d7d6255f3dfc/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Fx86-turing%2Fmain/log

I'll try to figure out what is different in this configuration

**Comment by ThomasRaoux** [comment]
So the problem is that when Vulkan is there the gpu dialects get registered by VulkanDialect but if it is not built it doesn't get registered. I added it as part of CUDA target in this 
https://github.com/google/iree/pull/6988 and that fixes the build.
